{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Day 2\n",
    "\n",
    "Today, we will start using nf-core pipelines to find differentially abundant genes in our dataset. \n",
    "We are using data from the following paper: https://www.nature.com/articles/s41593-023-01350-3#Sec10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "1. Please take some time to read through the paper and understand their approach, hypotheses and goals."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "What was the objective of the study?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The study's objective was to investigate the molecular/transcriptomic effects of oxycodon (or in generell opioid) withdrawal from mice suffering from peripheral nerve injury compared to control mice absent of chronic neuropathic pain in the  brain reward circuitry."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "What do the conditions mean?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "oxy: Treatment with oxycodon.\n",
    "\n",
    "\n",
    "sal: Treatment with saline, acting as control."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What do the genotypes mean?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SNI: Spared Nerve Injury Genotype (nerval removal) -> Chronic pair\n",
    "\n",
    "\n",
    "Sham: Control without special genotype (surgery without nerve cutting was performed, so e.g. skin pain is comparable) -> Surgery effects, but no chronic pain."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Imagine you are the bioinformatician in the group who conducted this study. They hand you the raw files and ask you to analyze them.\n",
    "\n",
    "What would you do?\n",
    "\n",
    "- Basic RNA Seq analysis workflow:\n",
    "- Quality control, adapter trimming, alignment against reference, post-analysis quality control etc.\n",
    "- Either do it manually, or look for pipelines for example in nf-core\n",
    "\n",
    "Which groups would you compare to each other?\n",
    "\n",
    "- SNI-oxy vs sham-sal\n",
    "- SNI-sal vs sham-sal\n",
    "- sham-oxy vs sham-sal\n",
    "- Compare all non-control genotypes + treatments to control (sham-sal) to eventually identify effects that are really based on the oxycodon withdrawal during chronic pain by ruling out effects that may arise from oxycodon treatment and withdrawal without condition itself or the SNI genotype.\n",
    "\n",
    "Please also mention which outcome you would expect to see from each comparison. \n",
    "\n",
    "- SNI-oxy vs sham-sal: Withdrawal in presenence of chronic pain \n",
    "- SNI-sal vs sham-sal: Disease only\n",
    "- Sham-oxy vs sham-sal: Withdrawal only"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Your group gave you a very suboptimal excel sheet (conditions_runs_oxy_project.xlsx) to get the information you need for each run they uploaded to the SRA.<br>\n",
    "So, instead of directly diving into downloading the data and starting the analysis, you first need to sort the lazy table.<br>\n",
    "Use Python and Pandas to get the table into a more sensible order.<br>\n",
    "Then, perform some overview analysis and plot the results\n",
    "1. How many samples do you have per condition?\n",
    "2. How many samples do you have per genotype?\n",
    "3. How often do you have each condition per genotype?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/8f/g9m13fjx2rjd968dsn9tzvcm0000gn/T/ipykernel_39961/72308387.py:6: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  df = df.fillna(False).replace(\"x\", True)\n",
      "/var/folders/8f/g9m13fjx2rjd968dsn9tzvcm0000gn/T/ipykernel_39961/72308387.py:6: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  df = df.fillna(False).replace(\"x\", True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient</th>\n",
       "      <th>Run</th>\n",
       "      <th>RNA-seq</th>\n",
       "      <th>DNA-seq</th>\n",
       "      <th>Condition</th>\n",
       "      <th>Genotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195508</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195509</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195516</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195517</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195506</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195511</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195514</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195519</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195505</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195510</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195513</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195518</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195507</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195512</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195515</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195520</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Patient          Run  RNA-seq  DNA-seq Condition Genotype\n",
       "3        ?  SRR23195508     True    False       Oxy      SNI\n",
       "4        ?  SRR23195509     True    False       Oxy      SNI\n",
       "11       ?  SRR23195516     True    False       Oxy      SNI\n",
       "12       ?  SRR23195517     True    False       Oxy      SNI\n",
       "1        ?  SRR23195506     True    False       Oxy     Sham\n",
       "6        ?  SRR23195511     True    False       Oxy     Sham\n",
       "9        ?  SRR23195514     True    False       Oxy     Sham\n",
       "14       ?  SRR23195519     True    False       Oxy     Sham\n",
       "0        ?  SRR23195505     True    False       Sal      SNI\n",
       "5        ?  SRR23195510     True    False       Sal      SNI\n",
       "8        ?  SRR23195513     True    False       Sal      SNI\n",
       "13       ?  SRR23195518     True    False       Sal      SNI\n",
       "2        ?  SRR23195507     True    False       Sal     Sham\n",
       "7        ?  SRR23195512     True    False       Sal     Sham\n",
       "10       ?  SRR23195515     True    False       Sal     Sham\n",
       "15       ?  SRR23195520     True    False       Sal     Sham"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# Import and sort fancy table\n",
    "df = pd.read_csv(\"/Users/nilswaffenschmidt/waffenschmidt/notebooks/day_02/conditions_runs_oxy_project.csv\", sep=\";\")\n",
    "df = df.fillna(False).replace(\"x\", True)\n",
    "\n",
    "# Wewrite conditions\n",
    "coniditons = [\"Sal\", \"Oxy\"]\n",
    "genotypes  = [\"SNI\", \"Sham\"]\n",
    "df[\"Condition\"] = np.select(df[[\"condition: Sal\", \"Condition: Oxy\"]].to_numpy().T, coniditons, default=None)\n",
    "df[\"Genotype\"]  = np.select(df[[\"Genotype: SNI\", \"Genotype: Sham\"]].to_numpy().T, genotypes, default=None)\n",
    "df = df.drop(['condition: Sal', 'Condition: Oxy', 'Genotype: SNI', 'Genotype: Sham'], axis=1)\n",
    "\n",
    "# Sort by Conditions and Genotypes\n",
    "df = df.sort_values(by=['Condition', 'Genotype'])\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have 8 samples per condition. 8 samples per genotype. 4 samples per condition per genotype."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "They were so kind to also provide you with the information of the number of bases per run, so that you can know how much space the data will take on your Cluster.<br>\n",
    "Add a new column to your fancy table with this information (base_counts.csv) and sort your dataframe according to this information and the condition.\n",
    "\n",
    "Then select the 2 smallest runs from your dataset and download them from SRA (maybe an nf-core pipeline can help here?...)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient</th>\n",
       "      <th>Run</th>\n",
       "      <th>RNA-seq</th>\n",
       "      <th>DNA-seq</th>\n",
       "      <th>Condition</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Bases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195515</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "      <td>8169101700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195513</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "      <td>8099181600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195507</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "      <td>8063298900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195518</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "      <td>7908500400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195506</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "      <td>7859530800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195520</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "      <td>7858146000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195512</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>Sham</td>\n",
       "      <td>7462857900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195510</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "      <td>7377388500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195514</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "      <td>7226808600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195509</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "      <td>7003550100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195519</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "      <td>6996050100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195508</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "      <td>6927786900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195505</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Sal</td>\n",
       "      <td>SNI</td>\n",
       "      <td>6922564500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195517</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "      <td>6863840400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195511</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>Sham</td>\n",
       "      <td>6456390900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>?</td>\n",
       "      <td>SRR23195516</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Oxy</td>\n",
       "      <td>SNI</td>\n",
       "      <td>6203117700</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Patient          Run  RNA-seq  DNA-seq Condition Genotype       Bases\n",
       "14       ?  SRR23195515     True    False       Sal     Sham  8169101700\n",
       "10       ?  SRR23195513     True    False       Sal      SNI  8099181600\n",
       "12       ?  SRR23195507     True    False       Sal     Sham  8063298900\n",
       "11       ?  SRR23195518     True    False       Sal      SNI  7908500400\n",
       "4        ?  SRR23195506     True    False       Oxy     Sham  7859530800\n",
       "15       ?  SRR23195520     True    False       Sal     Sham  7858146000\n",
       "13       ?  SRR23195512     True    False       Sal     Sham  7462857900\n",
       "9        ?  SRR23195510     True    False       Sal      SNI  7377388500\n",
       "6        ?  SRR23195514     True    False       Oxy     Sham  7226808600\n",
       "1        ?  SRR23195509     True    False       Oxy      SNI  7003550100\n",
       "7        ?  SRR23195519     True    False       Oxy     Sham  6996050100\n",
       "0        ?  SRR23195508     True    False       Oxy      SNI  6927786900\n",
       "8        ?  SRR23195505     True    False       Sal      SNI  6922564500\n",
       "3        ?  SRR23195517     True    False       Oxy      SNI  6863840400\n",
       "5        ?  SRR23195511     True    False       Oxy     Sham  6456390900\n",
       "2        ?  SRR23195516     True    False       Oxy      SNI  6203117700"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add Bases to fancy table\n",
    "df_counts = pd.read_csv(\"/Users/nilswaffenschmidt/waffenschmidt/notebooks/day_02/base_counts.csv\", sep=\",\")\n",
    "df_final  = pd.merge(left=df, right=df_counts, on=\"Run\")\n",
    "df_final  = df_final.sort_values(by=\"Bases\", ascending=True)\n",
    "df_final  = df_final.sort_values(['Bases'],ascending=False)\n",
    "\n",
    "df_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m\u001b[38;5;232m\u001b[48;5;43m N E X T F L O W \u001b[0;2m  ~  \u001b[mversion 25.04.7\u001b[m\n",
      "\u001b[K\n",
      "Pulling nf-core/fetchngs ...\n",
      "WARN: Cannot read project manifest -- Cause: API rate limit exceeded -- Provide your GitHub user name and password to get a higher rate limit\n",
      "API rate limit exceeded -- Provide your GitHub user name and password to get a higher rate limit\n"
     ]
    }
   ],
   "source": [
    "!nextflow run nf-core/fetchngs --input /Users/nilswaffenschmidt/waffenschmidt/notebooks/day_02/ids.csv -profile docker,arm --max_memory \"4GB\" --outdir /Users/nilswaffenschmidt/waffenschmidt/notebooks/day_02/nf_output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While your files are downloading, get back to the paper and explain how you would try to reproduce the analysis.<br>\n",
    "When you are done with this shout, so we can discuss the different ideas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "compWorkflows",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
